In [3]:
import pandas as pd


In [4]:
# Import the CSV file – NSMES1988new.csv into a dataframe
df = pd.read_csv("NSMES1988.csv")
df.head()

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


In [5]:
# Q2 Perform memory analysis of the new dataframe and compare it with the memory of the dataframe 
# in the previous week and mark your comments. 

df.memory_usage(deep=True)

Index            132
Unnamed: 0     35248
visits         35248
nvisits        35248
ovisits        35248
novisits       35248
emergency      35248
hospital       35248
health        245760
chronic        35248
adl           243229
region        242788
age            35248
gender        238774
married       227112
school         35248
income         35248
employed      225161
insurance     228127
medicaid      225108
dtype: int64

In [6]:
# df.describe
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   str    
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   str    
 10  region      4406 non-null   str    
 11  age         4406 non-null   float64
 12  gender      4406 non-null   str    
 13  married     4406 non-null   str    
 14  school      4406 non-null   int64  
 15  income      4406 non-null   float64
 16  employed    4406 non-null   str    
 17  insurance   4406 non-null   str    
 18  medicaid    4406 non-null   str    
dtypes: float64(2), int64(9), str(8)
memory

In [7]:
# Q3 Perform the following operations on age and income columns. Multiply age by 10 and income by 10000
df['age'] = df['age'] * 10
 

In [ ]:
df['income'] = df['income'] * 10000


In [9]:
df[['age', 'income']].head()

,age,income
0,69.0,2881.0
1,74.0,2747.8
2,66.0,653.2
3,76.0,658.8
4,79.0,658.8


Q4 Perform basic statistical analysis on the new dataframe and generate a brief report on 
the outcome. Save the dataframe as NSMES1988updated.csv file in the local space for 
possible future use.

In [ ]:
# Step 1: Sanity check that the Q3 transformation was actually written into the dataframe.
# NOTE: the two Q3 cells above are not repeatable - running them twice would scale the
# values a second time. Always use "Restart Kernel and Run All Cells" for this notebook.
print('age    :', df['age'].min(), 'to', df['age'].max(), '-> should be 66 to 109 years')
print('income :', df['income'].min(), 'to', df['income'].max(), '-> should be in dollars')

In [11]:
# Step 2: Basic statistical analysis of the numeric columns
# count, mean, std, min, quartiles and max
df.describe()

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,chronic,age,school,income
count,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000,4406.000000
mean,2203.500000,5.774399,1.618021,0.750794,0.536087,0.263504,0.295960,1.541988,74.024058,10.290286,2527.132047
std,1272.046972,6.759225,5.317056,3.652759,3.879506,0.703659,0.746398,1.349632,6.334050,3.738736,2924.647618
min,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,66.000000,0.000000,-1012.500000
25%,1102.250000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,69.000000,8.000000,912.150000
50%,2203.500000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,73.000000,11.000000,1698.150000
75%,3304.750000,8.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000,78.000000,12.000000,3172.850000
max,4406.000000,89.000000,104.000000,141.000000,155.000000,12.000000,8.000000,8.000000,109.000000,18.000000,54835.100000


In [9]:
# Step 3: Statistical summary of the categorical (text) columns
# count, number of unique categories, most frequent category and its frequency
df.describe(exclude='number')

,health,adl,region,gender,married,employed,insurance,medicaid
count,4406,4406,4406,4406,4406,4406,4406,4406
unique,3,2,4,2,2,2,2,2
top,average,normal,other,female,yes,no,yes,no
freq,3509,3507,1614,2628,2406,3951,3421,4004


In [12]:
# Step 4: A few extra measures - central tendency, spread and shape of the distribution
key_cols = ['visits', 'hospital', 'emergency', 'chronic', 'age', 'school', 'income']

stats = pd.DataFrame({
    'mean'  : df[key_cols].mean(),
    'median': df[key_cols].median(),
    'mode'  : df[key_cols].mode().iloc[0],
    'std'   : df[key_cols].std(),
    'var'   : df[key_cols].var(),
    'skew'  : df[key_cols].skew(),      # > 0 means right (positively) skewed
    'range' : df[key_cols].max() - df[key_cols].min()
}).round(2)

stats

,mean,median,mode,std,var,skew,range
visits,5.77,4.00,0.0,6.76,45.69,3.34,89.0
hospital,0.30,0.00,0.0,0.75,0.56,3.97,8.0
emergency,0.26,0.00,0.0,0.70,0.50,5.07,12.0
chronic,1.54,1.00,1.0,1.35,1.82,1.02,8.0
age,74.02,73.00,66.0,6.33,40.12,0.89,43.0
school,10.29,11.00,12.0,3.74,13.98,-0.44,18.0
income,2527.13,1698.15,432.0,2924.65,8553563.69,5.96,55847.6


In [13]:
# Step 5: Correlation between the numeric columns - which variables move together?
df[['visits', 'nvisits', 'hospital', 'emergency', 'chronic',
    'age', 'school', 'income']].corr().round(2)

,visits,nvisits,hospital,emergency,chronic,age,school,income
visits,1.00,0.23,0.24,0.16,0.26,0.00,0.06,-0.00
nvisits,0.23,1.00,0.05,0.05,0.04,-0.04,0.09,0.01
hospital,0.24,0.05,1.00,0.48,0.23,0.07,-0.04,-0.01
emergency,0.16,0.05,0.48,1.00,0.20,0.06,-0.07,-0.03
chronic,0.26,0.04,0.23,0.20,1.00,0.10,-0.07,-0.05
age,0.00,-0.04,0.07,0.06,0.10,1.00,-0.14,-0.07
school,0.06,0.09,-0.04,-0.07,-0.07,-0.14,1.00,0.26
income,-0.00,0.01,-0.01,-0.03,-0.05,-0.07,0.26,1.00


## Brief Report on the Statistical Analysis

**Dataset:** NSMES1988 (updated) — 4,406 records, 19 columns, no missing values in any column.

**1. Healthcare utilisation (visits, hospital, emergency, nvisits, ovisits)**
- `visits` (physician office visits) averages **5.77** but the median is only **4**, and the maximum is **89**. Skewness of **+3.34** confirms a strongly right-skewed distribution — most people visit the doctor a few times a year while a small group of heavy users pulls the mean up. For such a column the median describes a "typical" person better than the mean.
- `hospital` (mean 0.30) and `emergency` (mean 0.26) have a median and 75th percentile of **0**, i.e. more than three-quarters of the sample had no hospital stay and no emergency visit at all.

**2. Health status (chronic)**
- Number of chronic conditions averages **1.54** (median 1, max 8), so the average elderly respondent carries about one to two chronic conditions.

**3. Demographics after the Q3 transformation**
- `age` now reads in actual years: mean **74.0**, range **66 – 109**, mildly right-skewed (+0.89). The sample is entirely elderly (66+), as expected for a Medicare survey.
- `school` (years of education) averages **10.29** years, median 11, and is slightly left-skewed (−0.44) — most respondents completed high school or close to it.
- `income` now reads in dollars: mean **≈ $25,271**, median **≈ $16,982**, std **≈ $29,246**, range **−$10,125 to $548,351**. Skewness of **+5.96** shows a very heavy right tail — a few high-income respondents inflate the mean far above the median. The negative minimum is worth noting as a data-quality point (it represents a reported net loss rather than an error to be dropped blindly).

**4. Categorical variables**
- `health`: 3 categories, **average** dominates (3,509 of 4,406 ≈ 80%).
- `adl`: **normal** for 3,507 (≈ 80%); the rest have limited activities of daily living.
- `region`: 4 categories, **other** is the largest (1,614).
- `gender`: **female** majority (2,628 ≈ 60%); `married`: **yes** (2,406 ≈ 55%).
- `employed`: **no** for 3,951 (≈ 90%) — consistent with a retired population.
- `insurance`: **yes** for 3,421 (≈ 78%); `medicaid`: **no** for 4,004 (≈ 91%).

**5. Correlations**
- The strongest relationship is `hospital` ↔ `emergency` (**0.48**) — people admitted to hospital also use the emergency room more.
- `visits` correlates moderately with `chronic` (0.26), `hospital` (0.24) and `nvisits` (0.23): poorer health drives all forms of utilisation together.
- `school` ↔ `income` (**0.26**) is the only notable socio-economic link.
- `age` and `income` are essentially uncorrelated with `visits` (≈ 0.00), so utilisation in this elderly sample is explained by health condition rather than by age or income.

**Conclusion:** the dataframe is complete (no nulls) and describes an elderly, largely retired, mostly insured population in average self-reported health. The count-type utilisation columns and `income` are heavily right-skewed, so median/IQR should be preferred over mean/std for them, and any future modelling should use methods suited to skewed count data. The updated dataframe is saved below as `NSMES1988updated.csv`.

In [14]:
# Step 6: Save the updated dataframe for future use
# index=False so that pandas does not add one more unnamed index column
df.to_csv('NSMES1988updated.csv', index=False)

# Verify that the file was written correctly by reading it back
check = pd.read_csv('NSMES1988updated.csv')
print('Saved shape :', check.shape)
check.head()

Saved shape : (4406, 19)


,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,69.0,male,yes,6,2881.0,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,74.0,female,yes,10,2747.8,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,66.0,female,no,10,653.2,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,76.0,male,yes,3,658.8,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,79.0,female,yes,6,658.8,no,yes,no


In [15]:
print(df.dtypes)

Unnamed: 0      int64
visits          int64
nvisits         int64
ovisits         int64
novisits        int64
emergency       int64
hospital        int64
health            str
chronic         int64
adl               str
region            str
age           float64
gender            str
married           str
school          int64
income        float64
employed          str
insurance         str
medicaid          str
dtype: object


# Q5 Indicate which of the columns are not eligible for statistical analysis and indicate possible datatype changes, and report.
- Make changes to the recommended file from the previous step, export it as a new `.csv` file for possible future use (Optional).
- Prepare a brief report and enter it in the mark-up cells of the JupyterLab Notebook.

In [16]:
# Step 1: Classify every column - is it eligible for statistical (numeric) analysis or not?
# "Eligible"     -> mean / std / correlation are meaningful
# "Not eligible" -> only counts and frequencies are meaningful (or the column carries no information)

review = {
    # column        nature                       eligible?              recommended dtype
    'Unnamed: 0': ('Row identifier (1..4406)',  'NOT eligible',        'drop the column'),
    'visits':     ('Discrete count',            'Eligible',            'int16'),
    'nvisits':    ('Discrete count',            'Eligible',            'int16'),
    'ovisits':    ('Discrete count',            'Eligible',            'int16'),
    'novisits':   ('Discrete count',            'Eligible',            'int16'),
    'emergency':  ('Discrete count',            'Eligible',            'int16'),
    'hospital':   ('Discrete count',            'Eligible',            'int16'),
    'health':     ('Ordinal category (3)',      'NOT eligible',        'ordered category'),
    'chronic':    ('Discrete count',            'Eligible',            'int16'),
    'adl':        ('Ordinal category (2)',      'NOT eligible',        'ordered category'),
    'region':     ('Nominal category (4)',      'NOT eligible',        'category'),
    'age':        ('Continuous, whole years',   'Eligible',            'int16'),
    'gender':     ('Nominal category (2)',      'NOT eligible',        'category'),
    'married':    ('Binary yes/no',             'NOT eligible as text','bool'),
    'school':     ('Discrete count (years)',    'Eligible',            'int16'),
    'income':     ('Continuous (dollars)',      'Eligible',            'float32'),
    'employed':   ('Binary yes/no',             'NOT eligible as text','bool'),
    'insurance':  ('Binary yes/no',             'NOT eligible as text','bool'),
    'medicaid':   ('Binary yes/no',             'NOT eligible as text','bool'),
}

eligibility = pd.DataFrame(
    [(c, str(df[c].dtype), n, e, r) for c, (n, e, r) in review.items()],
    columns=['column', 'current dtype', 'nature of data', 'statistical analysis', 'recommended dtype']
).set_index('column')

eligibility

,current dtype,nature of data,statistical analysis,recommended dtype
column,,,,
Unnamed: 0,int64,Row identifier (1..4406),NOT eligible,drop the column
visits,int64,Discrete count,Eligible,int16
nvisits,int64,Discrete count,Eligible,int16
ovisits,int64,Discrete count,Eligible,int16
novisits,int64,Discrete count,Eligible,int16
emergency,int64,Discrete count,Eligible,int16
hospital,int64,Discrete count,Eligible,int16
health,str,Ordinal category (3),NOT eligible,ordered category
chronic,int64,Discrete count,Eligible,int16


In [17]:
# Step 2: Why the count columns fit in int16 and age fits in int16 - check the actual ranges
# int8 holds -128..127, int16 holds -32768..32767
num_cols = ['visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
            'hospital', 'chronic', 'age', 'school', 'income']

pd.DataFrame({'min': df[num_cols].min(),
              'max': df[num_cols].max(),
              'has decimals': [(df[c] % 1 != 0).any() for c in num_cols]})

,min,max,has decimals
visits,0.0,89.0,False
nvisits,0.0,104.0,False
ovisits,0.0,141.0,False
novisits,0.0,155.0,False
emergency,0.0,12.0,False
hospital,0.0,8.0,False
chronic,0.0,8.0,False
age,66.0,109.0,False
school,0.0,18.0,False
income,-1012.5,54835.1,True


In [18]:
# Step 3: Apply the recommended changes on a copy of the dataframe
mem_before = df.memory_usage(deep=True).sum() / 1024      # KB, before any change

df_opt = df.copy()

# (a) drop the identifier column - it carries no information
df_opt = df_opt.drop(columns=['Unnamed: 0'])

# (b) counts and age -> int16
for col in ['visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
            'hospital', 'chronic', 'school', 'age']:
    df_opt[col] = df_opt[col].astype('int16')

# (c) income -> float32 (still far more precision than the data needs)
df_opt['income'] = df_opt['income'].astype('float32')

# (d) ordinal text -> ORDERED category, so comparisons and sorting work
df_opt['health'] = pd.Categorical(df_opt['health'],
                                  categories=['poor', 'average', 'excellent'], ordered=True)
df_opt['adl'] = pd.Categorical(df_opt['adl'],
                               categories=['normal', 'limited'], ordered=True)

# (e) nominal text -> category
for col in ['region', 'gender']:
    df_opt[col] = df_opt[col].astype('category')

# (f) yes/no text -> bool (True means yes)
for col in ['married', 'employed', 'insurance', 'medicaid']:
    df_opt[col] = (df_opt[col] == 'yes')

df_opt.dtypes

visits          int16
nvisits         int16
ovisits         int16
novisits        int16
emergency       int16
hospital        int16
health       category
chronic         int16
adl          category
region       category
age             int16
gender       category
married          bool
school          int16
income        float32
employed         bool
insurance        bool
medicaid         bool
dtype: object

In [19]:
# Step 4: Compare the memory before and after the datatype changes
mem_after = df_opt.memory_usage(deep=True).sum() / 1024    # KB

print(f'Before : {mem_before:10.2f} KB   ({df.shape[1]} columns)')
print(f'After  : {mem_after:10.2f} KB   ({df_opt.shape[1]} columns)')
print(f'Saved  : {mem_before - mem_after:10.2f} KB  = {100*(1 - mem_after/mem_before):.1f} % reduction')

pd.DataFrame({'before (bytes)': df.memory_usage(deep=True),
              'after (bytes)': df_opt.memory_usage(deep=True)}).fillna(0).astype('int64')

Before :    2210.86 KB   (19 columns)
After  :     130.01 KB   (18 columns)
Saved  :    2080.84 KB  = 94.1 % reduction


,before (bytes),after (bytes)
Index,132,132
Unnamed: 0,35248,0
adl,243229,4625
age,35248,8812
chronic,35248,8812
emergency,35248,8812
employed,225161,4406
gender,238774,4514
health,245760,4681
hospital,35248,8812


In [20]:
# Step 5: Confirm that nothing was lost - the statistics are the same after the conversion
comparison = pd.DataFrame({
    'mean before': df[['visits', 'chronic', 'age', 'school', 'income']].mean(),
    'mean after' : df_opt[['visits', 'chronic', 'age', 'school', 'income']].mean(),
}).round(2)
print(comparison)

# The columns that are NOT eligible for numeric statistics are analysed with frequencies
for col in ['health', 'adl', 'region', 'gender', 'married', 'employed', 'insurance', 'medicaid']:
    print(f'\n--- {col} ---')
    print(df_opt[col].value_counts())

         mean before  mean after
visits          5.77        5.77
chronic         1.54        1.54
age            74.02       74.02
school         10.29       10.29
income       2527.13     2527.13

--- health ---
health
average      3509
poor          554
excellent     343
Name: count, dtype: int64

--- adl ---
adl
normal     3507
limited     899
Name: count, dtype: int64

--- region ---
region
other        1614
midwest      1157
northeast     837
west          798
Name: count, dtype: int64

--- gender ---
gender
female    2628
male      1778
Name: count, dtype: int64

--- married ---
married
True     2406
False    2000
Name: count, dtype: int64

--- employed ---
employed
False    3951
True      455
Name: count, dtype: int64

--- insurance ---
insurance
True     3421
False     985
Name: count, dtype: int64

--- medicaid ---
medicaid
False    4004
True      402
Name: count, dtype: int64


In [ ]:
# Step 6 (optional part of the question): export the optimised dataframe as a NEW csv file
df_opt.to_csv('NSMES1988optimized.csv', index=False)

# IMPORTANT: a .csv file stores plain text only - it does NOT remember the dtypes.
# So when this file is read back, the dtype dictionary has to be supplied again.
dtype_map = {**{c: 'int16' for c in ['visits', 'nvisits', 'ovisits', 'novisits',
                                     'emergency', 'hospital', 'chronic', 'school', 'age']},
             'income': 'float32',
             **{c: 'category' for c in ['health', 'adl', 'region', 'gender']},
             **{c: 'bool' for c in ['married', 'employed', 'insurance', 'medicaid']}}

reloaded = pd.read_csv('NSMES1988optimized.csv', dtype=dtype_map)
print('Shape          :', reloaded.shape)
print('Memory reloaded: %.2f KB' % (reloaded.memory_usage(deep=True).sum() / 1024))
reloaded.head()

Shape          : (4406, 18)
Memory reloaded: 130.29 KB


,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,5,0,0,0,0,1,average,2,normal,other,69,male,True,6,2881.000000,True,True,False
1,1,0,2,0,2,0,average,2,normal,other,74,female,True,10,2747.800049,False,True,False
2,13,0,0,0,3,3,poor,4,limited,other,66,female,False,10,653.200012,False,False,True
3,16,0,5,0,1,1,poor,2,limited,other,76,male,True,3,658.799988,False,True,False
4,3,0,0,0,0,0,average,2,limited,other,79,female,True,6,658.799988,False,True,False


: 